### PR TITLE
If the file contains backslashes, it will create an invalid file URL.

### PR DESCRIPTION
**Pull request by dmurdoch**
This issue https://github.com/dmurdoch/rgl/issues/391 was due to `webshot2::webshot` receiving `tempfile(ext = ".html")` as the `url` argument.  It recognized it as a file, and called `file_url(url)`, but didn't fix the backslashes, so it created an invalid URL.  This PR fixes that problem.

### CODE REVIEW

**Comment by wch** [comment]
@dmurdoch Thanks for the PR. Can you confirm it works with all forms of filenames that one can enter in Windows? I don't use Windows often so I'm not sure about this, but I think the following are some of the possibilities: `c:/a/b.html`, `c:\a\b.html`, `/a/b.html`, and `\\a\b.html`.

The reason I'm asking for testing is because the code originally had a conditional where it did _not_ set `winslash` on Windows. It came up as a topic of discussion in this PR and I couldn't remember why it was originally that way:

[#36 
](https://github.com/rstudio/webshot2/pull/36/files)

https://github.com/rstudio/webshot2/pull/36#discussion_r1290648192

**Comment by dmurdoch** [comment]
Sorry, I don't have access to Windows at all these days.  The only tests I saw are the ones described in that rgl issue (https://github.com/dmurdoch/rgl/issues/391).

Maybe `tempfile()` or `normalizePath()` has changed since #36?

**Comment by wch** [comment]
@cderv Does this fix look reasonable to you?

**Review by cderv — APPROVED** [review]
Yes I believe this is reasonable, especially because `normalizePath` is used here to create a FILE URI, and not just a regular windows path. File URI are supposed to be `/` even on windows. (https://en.wikipedia.org/wiki/File_URI_scheme). So all good to me!

FWIW In more general case, we also use `/` on windows in R Markdown  ecosystem, as we use `xfun::normalize_path()` which is a wrapper with other default, and symlink handling. 
https://github.com/yihui/xfun/blob/97c0043cfc177cba02fe582dc411959cb1f3699d/R/paths.R#L75

> some of the possibilities: c:/a/b.html, c:\a\b.html, /a/b.html, and \\a\b.html.

I believe this is not just a matter of forward slash. It is possible that `file:///` only work with local file anyway on a letter Drive. For UNC path (`\\server\path`) it shoud be `file://server` or maybe `file:////server/` 
but this is quite rare anyway.

**Comment by cderv** [comment]
Though I looked at the issue that triggers this (https://github.com/dmurdoch/rgl/issues/391). This is what I can see in the error message shared (https://github.com/dmurdoch/rgl/issues/391#issuecomment-1732398091) and I can reproduce locally with CRAN version of package. 
````
file:///C:\Users\cholbert\AppData\Local\Temp\RtmpYtOniO\file43c44fbe5168.htmlC:\Users\cholbert\AppData\Local\Temp\RtmpYtOniO\file43c44fbe5168.html screenshot completed
Warning messages:
1: In onRejected(...) :
An error occurred: Error in onFulfilled(value): Unable to find any visible elements for selectors.
````

As you can see the path is duplicated. I don't think this is a problem of normalize path changing to `/` in webshot2 because it is after what is causing the issue
https://github.com/rstudio/webshot2/blob/1e1edc7df78dd62ee5c9b8218b540097dee72d22/R/utils.R#L75-L81

The file (a temporary file in your case) is modified to add a anchor suffix at the end. However, the `anchor_suffix` is not correctly match, as when `\\` in the path, it will be equivalent to full path. 

This is on my Windows, and I believe this is the root of the issue
``` r
filename <- tempfile(fileext = ".html")
filename
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\Rtmp8KkAj1\\file843c532a63.html"
file <- sub("(.*)#[^#]*", "\\1", filename)
file
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\Rtmp8KkAj1\\file843c532a63.html"
anchor_suffix <- sub(file, "", filename)
anchor_suffix
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\Rtmp8KkAj1\\file843c532a63.html"
```

In `anchor_suffix <- sub(file, "", filename)`, `file` is supposed to be a regex pattern. And in a regex pattern `\\` means `\`, but it needs to be matched with `\\` in `filename`. So there is no replacement, and `anchor_suffix` returns the initial file path, which leads to duplication. 

By changing to `/` in your tempfile creation at https://github.com/dmurdoch/rgl/pull/392, you fixed the issue you had with **webshot2**
``` r
filename <- tempfile(fileext = ".html")
filename
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\RtmpMvplvi\\file72306ce8937.html"
filename <- normalizePath(filename, winslash = "/", mustWork = FALSE)
filename
#> [1] "C:/Users/chris/AppData/Local/Temp/RtmpMvplvi/file72306ce8937.html"
file <- sub("(.*)#[^#]*", "\\1", filename)
file
#> [1] "C:/Users/chris/AppData/Local/Temp/RtmpMvplvi/file72306ce8937.html"
anchor_suffix <- sub(file, "", filename)
anchor_suffix
#> [1] ""
```

So either should **webshot2** do a first normalization without checking existant, as it seems to handle anchor
``` r
filename <- tempfile(fileext = ".html")
# add anchor
filename <- paste0(filename, "#2")
filename
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\RtmpIHZ5A8\\file91c4298f4de8.html#2"
# normalize
filename <- normalizePath(filename, winslash = "/", mustWork = FALSE)
filename
#> [1] "C:/Users/chris/AppData/Local/Temp/RtmpIHZ5A8/file91c4298f4de8.html#2"
```

Or probably simpler is to set `fixed = TRUE` in `sub()`
``` r
filename <- tempfile(fileext = ".html")
filename
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\Rtmpkn2f5D\\file515045c25f5f.html"
file <- sub("(.*)#[^#]*", "\\1", filename)
file
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\Rtmpkn2f5D\\file515045c25f5f.html"
anchor_suffix <- sub(file, "", filename, fixed = TRUE)
anchor_suffix
#> [1] ""
```

``` r
filename <- tempfile(fileext = ".html")
filename
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\RtmpYNDYy4\\file2a8840ad4c10.html"
filename <- paste0(filename, "#2")
filename
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\RtmpYNDYy4\\file2a8840ad4c10.html#2"
file <- sub("(.*)#[^#]*", "\\1", filename)
file
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\RtmpYNDYy4\\file2a8840ad4c10.html"
anchor_suffix <- sub(file, "", filename, fixed = TRUE)
anchor_suffix
#> [1] "#2"
```

And this will surely help avoid this type of issue in the future 😉 
* https://github.com/rstudio/webshot2/issues/60

File: R/utils.R [hunk 1]
```diff
@@ -76,7 +76,7 @@ file_url <- function(filename) {
   anchor_suffix <- sub(file, "", filename)
```
Comment (cderv) [inline R/utils.R]: See my main comment, but I believe this will solve your issue in **rgl**

```suggestion
  anchor_suffix <- sub(file, "", filename, fixed = TRUE)
```

File: R/utils.R [hunk 2]
```diff
@@ -76,7 +76,7 @@ file_url <- function(filename) {
...
   enc2(paste0(
     "file:///",
-    normalizePath(path = file, mustWork = TRUE),
+    normalizePath(path = file, mustWork = TRUE, winslash = "/"),
```
Comment (cderv) [inline R/utils.R:79]: This is still useful to do IMO on windows, but it won't be a problem for the issue I think as Windows handles the conversion. 

Probably worth testing... Ping me when PR is updated, and I'll try the PR of webshot2 with rgl on windows

**Review by cderv — COMMENTED** [review]
## TL;DR

Setting `fixed = TRUE` to `sub()` for saving anchor part should solve the real issue here. This is regex pattern issue - always tricky to use filepath as regex pattern because of escaping of regex

## Details

Though I looked at the issue that triggers this (https://github.com/dmurdoch/rgl/issues/391). This is what I can see in the error message shared (https://github.com/dmurdoch/rgl/issues/391#issuecomment-1732398091) and I can reproduce locally with CRAN version of package. 
````
file:///C:\Users\cholbert\AppData\Local\Temp\RtmpYtOniO\file43c44fbe5168.htmlC:\Users\cholbert\AppData\Local\Temp\RtmpYtOniO\file43c44fbe5168.html screenshot completed
Warning messages:
1: In onRejected(...) :
An error occurred: Error in onFulfilled(value): Unable to find any visible elements for selectors.
````

As you can see the path is duplicated. I don't think this is a problem of normalize path changing to `/` in webshot2 because it is after what is causing the issue
https://github.com/rstudio/webshot2/blob/1e1edc7df78dd62ee5c9b8218b540097dee72d22/R/utils.R#L75-L81

The file (a temporary file in your case) is modified to add a anchor suffix at the end. However, the `anchor_suffix` is not correctly match, as when `\\` in the path, it will be equivalent to full path. 

This is on my Windows, and I believe this is the root of the issue
``` r
filename <- tempfile(fileext = ".html")
filename
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\Rtmp8KkAj1\\file843c532a63.html"
file <- sub("(.*)#[^#]*", "\\1", filename)
file
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\Rtmp8KkAj1\\file843c532a63.html"
anchor_suffix <- sub(file, "", filename)
anchor_suffix
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\Rtmp8KkAj1\\file843c532a63.html"
```

In `anchor_suffix <- sub(file, "", filename)`, `file` is supposed to be a regex pattern. And in a regex pattern `\\` means `\`, but it needs to be matched with `\\` in `filename`. So there is no replacement, and `anchor_suffix` returns the initial file path, which leads to duplication. 

By changing to `/` in your tempfile creation at https://github.com/dmurdoch/rgl/pull/392, you fixed the issue you had with **webshot2**
``` r
filename <- tempfile(fileext = ".html")
filename
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\RtmpMvplvi\\file72306ce8937.html"
filename <- normalizePath(filename, winslash = "/", mustWork = FALSE)
filename
#> [1] "C:/Users/chris/AppData/Local/Temp/RtmpMvplvi/file72306ce8937.html"
file <- sub("(.*)#[^#]*", "\\1", filename)
file
#> [1] "C:/Users/chris/AppData/Local/Temp/RtmpMvplvi/file72306ce8937.html"
anchor_suffix <- sub(file, "", filename)
anchor_suffix
#> [1] ""
```

So either should **webshot2** do a first normalization without checking existant, as it seems to handle anchor
``` r
filename <- tempfile(fileext = ".html")
# add anchor
filename <- paste0(filename, "#2")
filename
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\RtmpIHZ5A8\\file91c4298f4de8.html#2"
# normalize
filename <- normalizePath(filename, winslash = "/", mustWork = FALSE)
filename
#> [1] "C:/Users/chris/AppData/Local/Temp/RtmpIHZ5A8/file91c4298f4de8.html#2"
```

Or probably simpler is to set `fixed = TRUE` in `sub()`
``` r
filename <- tempfile(fileext = ".html")
filename
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\Rtmpkn2f5D\\file515045c25f5f.html"
file <- sub("(.*)#[^#]*", "\\1", filename)
file
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\Rtmpkn2f5D\\file515045c25f5f.html"
anchor_suffix <- sub(file, "", filename, fixed = TRUE)
anchor_suffix
#> [1] ""
```

``` r
filename <- tempfile(fileext = ".html")
filename
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\RtmpYNDYy4\\file2a8840ad4c10.html"
filename <- paste0(filename, "#2")
filename
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\RtmpYNDYy4\\file2a8840ad4c10.html#2"
file <- sub("(.*)#[^#]*", "\\1", filename)
file
#> [1] "C:\\Users\\chris\\AppData\\Local\\Temp\\RtmpYNDYy4\\file2a8840ad4c10.html"
anchor_suffix <- sub(file, "", filename, fixed = TRUE)
anchor_suffix
#> [1] "#2"
```

And this will surely help avoid this type of issue in the future 😉 
* https://github.com/rstudio/webshot2/issues/60

Hope it helps

And @wch, instead of using `file://` URI and needed to build URL, another way could be to only rely on a URL by setting up a local server also for file ? We do that in `pagedown::chrome_print()` to load everything in browser before doing the screenshot (required for paged.js): https://github.com/rstudio/pagedown/blob/466c1c1e8fc4a679aeff25bdd19fd834c0b78bbd/R/chrome.R#L148-L169
Though it would require to handle local `servr` management, especially closing when done.

**Comment by dmurdoch** [comment]
@cderv :  Thanks for noticing that.  I've added your suggested change, and all the checks have passed.

**Comment by cderv** [comment]
Thanks @dmurdoch ! I believe with this PR you don't need you workaround in **rgl**. 
You could check this is fixed for you by using this PR version of **webshot2** and unpatched version of **rgl**

Did you check it already ?

**Comment by wch** [comment]
> And @wch, instead of using `file://` URI and needed to build URL, another way could be to only rely on a URL by setting up a local server also for file ? We do that in `pagedown::chrome_print()` to load everything in browser before doing the screenshot (required for paged.js): [rstudio/pagedown@`466c1c1`/R/chrome.R#L148-L169](https://github.com/rstudio/pagedown/blob/466c1c1e8fc4a679aeff25bdd19fd834c0b78bbd/R/chrome.R#L148-L169) Though it would require to handle local `servr` management, especially closing when done.

I think it would be good to add this functionality, although I don't really like the idea of taking a hard dependency on httpuv. Note that in httpuv, we recently added a function called `runStaticServer()` which does basically the same thing as servr. https://github.com/rstudio/httpuv/pull/380

**Comment by wch** [comment]
Thanks for the thorough look at this PR!

**Comment by cderv** [comment]
> Note that in httpuv, we recently added a function called runStaticServer() which does basically the same thing as servr. 

Oh good to know ! 
And yes, I understand the hard dependency. It is also easy enough for a user to run the server before doing the webshot probably.
